ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Validation.hs

```diff
@@ -0,0 +1,414 @@
+      -- the reconstructed index (truncate first).
+      primaryIndex            <- reconstructPrimaryIndex epochInfo hashInfo
+        shouldBeFinalised (map Secondary.blockOrEBB entries)
+      primaryIndexFileExists  <- doesFileExist primaryIndexFile
+      primaryIndexFileMatches <- if primaryIndexFileExists
+        then EH.try errLoad (Primary.load hasFS err epoch) >>= \case
+          Left _                     -> do
+            traceWith tracer $ InvalidPrimaryIndex epoch
+            return False
+          Right primaryIndexFromFile ->
+            return $ primaryIndexFromFile == primaryIndex
+        else do
+          traceWith tracer $ MissingPrimaryIndex epoch
+          return False
+      unless primaryIndexFileMatches $ do
+        traceWith tracer $ RewritePrimaryIndex epoch
+        Primary.write hasFS epoch primaryIndex
+
+      return $  (\entry -> (Secondary.blockOrEBB entry, Secondary.headerHash entry))
+            <$> lastMaybe entries
+  where
+    epochFile          = renderFile "epoch"     epoch
+    primaryIndexFile   = renderFile "primary"   epoch
+    secondaryIndexFile = renderFile "secondary" epoch
+
+    HasFS { hTruncate, doesFileExist } = hasFS
+
+    trace = lift . traceWith tracer
+
+    -- | Handle only 'InvalidFileError', which is the only error that can be
+    -- thrown while load a primary or a secondary index file
+    errLoad :: ErrorHandling UnexpectedError m
+    errLoad = EH.embed UnexpectedError
+      (\case
+        UnexpectedError (e@InvalidFileError {}) -> Just e
+        _ -> Nothing) err
+
+-- | Reconstruct a 'PrimaryIndex' based on a list of 'Secondary.Entry's.
+reconstructPrimaryIndex
+  :: forall m hash. Monad m
+  => EpochInfo m
+  -> HashInfo hash
+  -> ShouldBeFinalised
+  -> [BlockOrEBB]
+  -> m PrimaryIndex
+reconstructPrimaryIndex epochInfo HashInfo { hashSize } shouldBeFinalised
+                        blockOrEBBs = do
+    relSlots <- mapM toRelativeSlot blockOrEBBs
+    let secondaryOffsets = 0 : go 0 0 relSlots
+
+    -- This can only fail if the slot numbers of the entries are not
+    -- monotonically increasing.
+    return $ fromMaybe (error msg) $ Primary.mk secondaryOffsets
+  where
+    msg = "blocks have non-increasing slot numbers"
+
+    toRelativeSlot :: BlockOrEBB -> m RelativeSlot
+    toRelativeSlot (EBB _)      = return 0
+    toRelativeSlot (Block slot) = _relativeSlot <$>
+      epochInfoBlockRelative epochInfo slot
+
+    go
+      :: HasCallStack
+      => RelativeSlot
+      -> SecondaryOffset
+      -> [RelativeSlot]
+      -> [SecondaryOffset]
+    go nextExpectedRelSlot lastSecondaryOffset = \case
+      [] -> case shouldBeFinalised of
+        ShouldNotBeFinalised        -> []
+        ShouldBeFinalised epochSize ->
+            Primary.backfill (succ finalRelSlot) nextExpectedRelSlot
+              lastSecondaryOffset
+          where
+            finalRelSlot = maxRelativeSlot epochSize
```

Comment:
Idea: introduce a variant of `Primary.backfill` that takes an `EpochSize` as argument.

Use it in `startNewEpoch` too.